### PR TITLE
Ns/plug zk v2

### PR DESCRIPTION
**Pull request by nsarlin-zama**
closes: https://github.com/zama-ai/tfhe-rs-internal/issues/772
### PR content/description
Plugs ZK V2. The CompactPkeCrs type is now an enum with a variant for V1 and V2, each holding the right 'PublicParam' type from `tfhe_zk_pok`. The proof is similarly wrapped in an enum. A prove method has been added on the Crs object that will generate a proof compatible with this Crs. Similarly, a verify method will always reject proofs that are not of the right kind.

Backward compatibility is maintained and old crs/proofs are automatically wrapped in the V1 scheme. A fix was needed in versionable for this, to make it handle type that evolve from `#[repr(transparent)]` to their own type.

By default new Crs are created using v2, but methods have been added to create a Crs with v1.

Proof conformance checks the version of the scheme.

### Check-list:

* [x] Tests for the changes have been added -> Need to add a v2 proof in data repo
* [x] Param update
* [x] Docs have been added / updated (for bug fixes / features)
* [x] Relevant issues are marked as resolved/closed, related issues are linked in the description
* [x] Check for breaking changes (including serialization changes) and add them to commit message following the conventional commit [specification][conventional-breaking]

[conventional-breaking]: https://www.conventionalcommits.org/en/v1.0.0/#commit-message-with-description-and-breaking-change-footer

### CODE REVIEW

File: tfhe/src/shortint/parameters/compact_public_key_only/p_fail_2_minus_64/ks_pbs.rs
Comment (IceTDrinker) [inline tfhe/src/shortint/parameters/compact_public_key_only/p_fail_2_minus_64/ks_pbs.rs:12]: I'll check if the keyswitching parameters need an update as well

Comment (IceTDrinker) [reply tfhe/src/shortint/parameters/compact_public_key_only/p_fail_2_minus_64/ks_pbs.rs:12]: looks like the keyswitch to big key needs a small update, I'll give you the info to update and have legacy constants there as well

**Review by mayeul-zama — COMMENTED** [review]
Just a few comments
FYI I only had a quick look at the "plug zk v2" commit

File: utils/tfhe-versionable/examples/transparent_then_not.rs
Comment (tmontaigu) [inline utils/tfhe-versionable/examples/transparent_then_not.rs:19]: That sentence confuses me, to me, the last version that was transparent is V1, not V0, so if I follow the `that starts at the last version that was transparent` the enum should start at V1

File: utils/tfhe-versionable/examples/transparent_then_not.rs
Comment (nsarlin-zama) [reply utils/tfhe-versionable/examples/transparent_then_not.rs:7]: Ok I will try to make the example more readable. I don't think this is just an ordering issue because I would like the order of the definitions the match the explanation in the comments

File: utils/tfhe-versionable/examples/transparent_then_not.rs
Comment (nsarlin-zama) [reply utils/tfhe-versionable/examples/transparent_then_not.rs:19]: Yeah I'm confused too, I will reword that

**Review by mayeul-zama — APPROVED** [review]
Thanks!

**Comment by IceTDrinker** [comment]
please don't merge right away, I'd like to have a chance to discuss the keyswitching parameters before you do (and read the PR as well)

File: utils/tfhe-versionable/examples/transparent_then_not.rs
Comment (tmontaigu) [inline utils/tfhe-versionable/examples/transparent_then_not.rs:26]: I think its better worded that way, but, wording now makes me understand that the enum should be
```rust
enum MyStructWrapperVersions<T> {
    V0(MyStructV0),
    V1(MyStrucV1<T>),
    V2(MyStructWrapper<T>),
}
```

as for V0 and V1 the Wrapper was transparent, and so the inner types were `MyStructV0` and `MyStrucV1`.

because as it stands,
```rust
enum MyStructWrapperVersions<T> {
    V0(MyStructWrapperV0),
    V1(MyStructWrapperV1<T>),
    V2(MyStructWrapper<T>),
}
```
Does not look different  from what is usally done for non transparent structs

Comment (nsarlin-zama) [reply utils/tfhe-versionable/examples/transparent_then_not.rs:26]: What I mean here is that in v0 and v1(versions of the app) the `MyStructWrapperVersions` enum does not exist, so we need to create it directly in v2 with at least the same versions as `MyStructVersions`.

Whereas with normal non transparent types you build your enum one version at a time, so you never create from scratch a dispatch enum with 3 variants

Comment (nsarlin-zama) [reply utils/tfhe-versionable/examples/transparent_then_not.rs:26]: the examples in this repo try to follow a pattern where "mod v0", "mod v1"... simulate previous versions of an application to show how the versioning evolve with time

Comment (nsarlin-zama) [reply utils/tfhe-versionable/examples/transparent_then_not.rs]: I reworded it a bit

File: tfhe/src/zk/mod.rs
Comment (IceTDrinker) [inline tfhe/src/zk/mod.rs:72]: new type ?

**Review by IceTDrinker — COMMENTED** [review]
It seems we are missing a check on k_max <= d it seems, otherwise we are looking good, after the parameter stuff

File: tfhe/src/shortint/parameters/compact_public_key_only/p_fail_2_minus_64/ks_pbs.rs
Comment (IceTDrinker) [inline tfhe/src/shortint/parameters/compact_public_key_only/p_fail_2_minus_64/ks_pbs.rs:23]: did you do the keyswitch parameters update ?

File: tfhe/src/shortint/parameters/compact_public_key_only/p_fail_2_minus_64/ks_pbs.rs
Comment (IceTDrinker) [inline tfhe/src/shortint/parameters/compact_public_key_only/p_fail_2_minus_64/ks_pbs.rs:22]: also this does not respect the potential TFHE-rs version being in the name of consts

File: tfhe/src/shortint/parameters/compact_public_key_only/p_fail_2_minus_64/ks_pbs.rs
Comment (nsarlin-zama) [reply tfhe/src/shortint/parameters/compact_public_key_only/p_fail_2_minus_64/ks_pbs.rs:23]: not yet !

File: tfhe/src/shortint/parameters/compact_public_key_only/p_fail_2_minus_64/ks_pbs.rs
Comment (nsarlin-zama) [reply tfhe/src/shortint/parameters/compact_public_key_only/p_fail_2_minus_64/ks_pbs.rs:22]: I know, I did not reworked the params yet

**Pull request merged**
